packages/lib/server/repository/profile.ts

```diff
@@ -0,0 +1,500 @@
+import type { User as PrismaUser } from "@prisma/client";
+import { v4 as uuidv4 } from "uuid";
+
+import { whereClauseForOrgWithSlugOrRequestedSlug } from "@calcom/ee/organizations/lib/orgDomains";
+import { safeStringify } from "@calcom/lib/safeStringify";
+import prisma from "@calcom/prisma";
+import type { Team } from "@calcom/prisma/client";
+import type { UpId, UserAsPersonalProfile, UserProfile } from "@calcom/types/UserProfile";
+
+import logger from "../../logger";
+import { getParsedTeam } from "./teamUtils";
+import { UserRepository } from "./user";
+
+const log = logger.getSubLogger({ prefix: ["repository/profile"] });
+const organizationSelect = {
+  id: true,
+  slug: true,
+  name: true,
+  metadata: true,
+  calVideoLogo: true,
+};
+
+export enum LookupTarget {
+  User,
+  Profile,
+}
+
+export class ProfileRepository {
+  static generateProfileUid() {
+    return uuidv4();
+  }
+
+  private static getInheritedDataFromUser({ user }: { user: PrismaUser }) {
+    return {
+      name: user.name,
+      avatarUrl: user.avatarUrl,
+      startTime: user.startTime,
+      endTime: user.endTime,
+      bufferTime: user.bufferTime,
+      avatar: user.avatar,
+    };
+  }
+
+  static getLookupTarget(upId: UpId) {
+    if (upId.startsWith("usr-")) {
+      return {
+        type: LookupTarget.User,
+        id: parseInt(upId.replace("usr-", "")),
+      } as const;
+    }
+    return {
+      type: LookupTarget.Profile,
+      id: parseInt(upId),
+    } as const;
+  }
+
+  private static async _create({
+    userId,
+    organizationId,
+    username,
+    email,
+    movedFromUserId,
+  }: {
+    userId: number;
+    organizationId: number;
+    username: string | null;
+    email: string;
+    movedFromUserId?: number;
+  }) {
+    log.debug("_create", safeStringify({ userId, organizationId, username, email }));
+    return prisma.profile.create({
+      data: {
+        uid: ProfileRepository.generateProfileUid(),
+        user: {
+          connect: {
+            id: userId,
+          },
+        },
+        organization: {
+          connect: {
+            id: organizationId,
+          },
+        },
+        ...(movedFromUserId
+          ? {
+              movedFromUser: {
+                connect: {
+                  id: movedFromUserId,
+                },
+              },
+            }
+          : null),
+
+        username: username || email.split("@")[0],
+      },
+    });
+  }
+
+  /**
+   * Accepts `email` as a source to derive username from when username is null
+   * @returns
+   */
+  static create({
+    userId,
+    organizationId,
+    username,
+    email,
+  }: {
+    userId: number;
+    organizationId: number;
+    username: string | null;
+    email: string;
+  }) {
+    return ProfileRepository._create({ userId, organizationId, username, email });
+  }
+
+  static async upsert({
```

Comment:
Actually _create is more generic and it can be called from **create** as well as **createForExistingUser** which are specialized functions. That's why I created _create, upsert didn't have such a requirement.